### PR TITLE
Management command to create a Tableau "Connected App"

### PR DESCRIPTION
**Pull request by AddisonDunn**
## Product Description

This is the first PR of a multi-part series on improving Tableau permissions. This management command adds a connection to Tableau called a [connected app](https://help.tableau.com/current/online/en-us/connected_apps.htm). This connection is designed for web applications embedding Tableau content and will enable us to use the [Tableau Embedding API](https://help.tableau.com/current/api/embedding_api/en-us/docs/embedding_api_web_authoring.html) for embedding views.

I want to replace our current way of embedding dashboards via Trusted Authentication with this API because it doesn't require someone to sign in as a Tableau user in order to see a Tableau view, which is the cause of many security/integration flaws that have popped up.

Instead, HQ will be trusted by Tableau Server as a "connected app". Permissions will look like they do with any other type of report (view vs. edit) and we can do other things that the BHA is starting to ask for, such as [filtering the data for a view](https://help.tableau.com/current/api/embedding_api/en-us/docs/embedding_api_filter.html) based on which clinic the user is a part of.

Jira ticket [USH-2330](https://dimagi-dev.atlassian.net/browse/USH-2230?atlOrigin=eyJpIjoiNmUwM2EzYjMzNGEyNGRkZDg5YmU2MDllNGVjZjY1MzEiLCJwIjoiaiJ9).

## Technical Summary

Once this management command is merged, we can create a Tableau "connected app" connection for the BHA domain, then as required for other CICT projects (or beyond). The secret that's stored in the TableauConnection model will be used in [USH-2231 ](https://dimagi-dev.atlassian.net/browse/USH-2231?atlOrigin=eyJpIjoiNWMyZWJjNGY3NjAzNDU4Y2EwNWQwOWIzZWIzNzU5YjUiLCJwIjoiaiJ9)to create [JWTs](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_auth.htm#make-a-sign-in-request-with-jwt) (another password-like string that's just used for sending requests) for authentication when embedding views via the new API.

If you have any other questions, like "what is a Personal Access Token?", let me know!

## Feature Flag

Technically this management command can be run on any domain, but practically should only be run on domains with the feature flag EMBEDDED_TABLEAU.

### Safety story

Tested locally, I think that's all that's required for now. There's no risk of breaking anything until USH-2231 when we modify code that actually affects the embedded Tableau report pages.

The highest risk is probably security related. The `secret_key` is a stored value that will be used to generate JWTs for authenticated requests. I'd love to hear if you have any hold-ups about storing that in Postgres. Otherwise, I think this is pretty secure.

### Automated test coverage

None.

### QA Plan

None.

### Migrations

- [X] The migrations in this code can be safely applied first independently of the code

Just adds a new model, so should be fine to run with deploy.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/reports/management/commands/create_tableau_connection.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+logger = logging.getLogger('create_tableau_connection')
+logger.setLevel('DEBUG')
+
+
+class RequestFailure(Exception):
+    pass
+
+
+TABLEAU_CONNECTED_APP_NAME = "commcarehq_embedding"
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/create_tableau_connection.py:18]: Should this be configurable? If multiple domains create connections does this stay the same?

File: corehq/apps/reports/management/commands/create_tableau_connection.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        request_url = f"{self.url_base}/sites/{self.site_id}/connected-applications"
+        response_body = self._process_response(
+            response=requests.get(request_url, headers=self.headers),
+            custom_msg_if_failure="Request to check if a connected app already exists failed."
+        )
+        if response_body["connectedApplications"]:
+            connected_apps = response_body["connectedApplications"]["connectedApplication"]
+            for app in connected_apps:
+                if app["name"] == "commcarehq_embedding":
+                    return (True, app["clientId"])
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/create_tableau_connection.py:164]: nit: tuple braces are unnecessary
```suggestion
                    return True, app["clientId"]
```

File: corehq/apps/reports/management/commands/create_tableau_connection.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        self.sign_in()
+        try:
+            # The site ID is necessary for making site-specific API requests.
+            self.site_id = self.get_site_id()
+            (connected_app_exists, connected_app_client_id) = self.check_if_connected_app_exists()
+            if connected_app_exists:
+                overwrite = input("\nA CommCareHQ connected app already exists on this site. Overwrite and "
+                                  "create a new one? [y/n]")
+                if overwrite == 'y':
+                    self.delete_existing_connected_app(connected_app_client_id)
```
Comment (snopoke) [inline corehq/apps/reports/management/commands/create_tableau_connection.py:93]: what if the app is for a different domain?

**Review by snopoke — COMMENTED** [review]
I have some questions about how the app is used. Will there need to be a different app per domain? Or is it a global app per Tableau Server? Depending on how that works I think we would need to be able to configure the app name and also record more details in the Django model to allow us to link the two.

Other than that the code is OK though I think it would be nice to extract the tableau API code into a separate class or module which would make the management command much cleaner.

File: corehq/apps/reports/management/commands/create_tableau_connection.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            self.create_connected_app()
+            logger.info("App succesffully created. Command execution completed.")
+            self.sign_out()
+        # I think it's ok to ignore the flake8 error for this bare except clause since we re-raise the exception.
+        except:  # noqa: E722
+            logger.info("Command execution interrupted.")
+            # Submit a sign out request on any interruption to ensure the token we were given upon sign in is no
+            # longer valid.
+            self.sign_out()
+            raise
```
Comment (esoergel) [inline corehq/apps/reports/management/commands/create_tableau_connection.py:107]: While I think this is fine as-is, this pattern would be a good place to try a `with` statement if you feeling like playing with context managers.

**Pull request closed without merging**